XiEditor/StyleMap.swift

```diff
@@ -61,15 +57,46 @@ func utf8_offset_to_utf16(_ s: String, _ ix: Int) -> Int {
 class StyleMap {
     private var map: [Style?] = []
 
-    func defStyle(json: [String: AnyObject]) {
+    func defStyle(for fromFont: NSFont, json: [String: AnyObject]) {
         guard let styleID = json["id"] as? Int else { return }
-        let fgColor = json["fg_color"] as? UInt32 ?? 0xFF000000
-        let bgColor = json["bg_color"] as? UInt32 ?? 0
-        let weight = json["weight"] as? UInt16 ?? 400
-        let underline = json["underline"] as? Bool ?? false
+
+        var attrs: [String: Any] = [
+            NSForegroundColorAttributeName: json["fg_color"] as? NSColor ?? NSColor.black,
+        ]
+
+        if let bgColor = json["bg_color"] as? NSColor, bgColor.alphaComponent != 0.0 {
+            attrs[NSBackgroundColorAttributeName] = bgColor
+        }
+
+        if let underline = json["underline"] as? Bool, underline {
+            attrs[NSUnderlineStyleAttributeName] = NSUnderlineStyle.styleSingle.rawValue
+        }
+
+        let fm = NSFontManager.shared()
+        var font: NSFont?
+
         let italic = json["italic"] as? Bool ?? false
+        // convert to NSFont weight: (100-500 -> 2-6 (5 normal weight), 600-800 -> 8-10, 900 -> 12
+        let w = Float(json["weight"] as? Int ?? fm.weight(of: fromFont))
+        let weight = Int(floor(1 + w * (0.01 + 3e-6 * w)))
```

Comment:
@raphlinus Updated to use the suggested formula. Hower, it currently maps `1000` to `14` (and not `15`).

If this is a issue, I could also update the logic to a select, e.g.

```swift
func fontWeight(_ input: Int) -> Int {
    switch input {
    case 100...500:
        return input / 100 + 1
    case 600...800:
        return input / 100 + 2
    case 900:
        return 12
    case 1000:
        return 15
    default:
        return 5
    }
}
```